ega-data-api-dataedge/src/main/java/htsjdk/samtools/seekablestream/EgaSeekableCachedResStream.java

```diff
@@ -0,0 +1,310 @@
+        long fsize = this.contentLength;
+        int bytesToRead = (int) Math.min(fsize - position, len);
+
+        int cachePage = (int) (position / PAGE_SIZE); // 0,1,2,...
+
+        try {
+            byte[] page = this.get(cachePage);
+
+            int page_offset = (int) (position - cachePage * PAGE_SIZE); // delta page start to 'Read from'
+            int bytesToCopy = Math.min(bytesToRead, page.length - page_offset); // don't read past end of page
+            int offset_ = offset;
+            System.arraycopy(page, page_offset, buffer, offset_, bytesToCopy);
+            offset_ += bytesToCopy;
+            this.position += bytesToCopy;
+
+            int bytesRemaining = bytesToRead - bytesToCopy;
+            while (bytesRemaining > 0) {
+                page = this.get(++cachePage); // this.cache.get(cachePage+1);
+                bytesToCopy = Math.min(bytesRemaining, page.length); // don't read past end of page
+                System.arraycopy(page, 0, buffer, offset_, bytesToCopy);
+                bytesRemaining -= bytesToCopy;
+                offset_ += bytesToCopy;
+                this.position += bytesToCopy;
+            }
+        } catch (ExecutionException e) {
+            System.out.println(e);
+            return 0;
+        }
+
+        return bytesToRead;
+    }
+
+
+    public void close() throws IOException {
+        // Nothing to do
+    }
+
+
+    public int read() throws IOException {
+        byte[] tmp = new byte[1];
+        read(tmp, 0, 1);
+        return (int) tmp[0] & 0xFF;
+    }
+
+    @Override
+    public String getSource() {
+        return hack ? null : url.toString() + hack_extension;
+    }
+
+    // A hack to return null as source, which will then default to BAM format in HTSJDK
+    public EgaSeekableCachedResStream setSourceNull(boolean hack) {
+        this.hack = hack;
+        return this;
+    }
+
+    // A hack to fool HTSJDK to recognize the file format based on the extension
+    public EgaSeekableCachedResStream setExtension(String hack_extension) {
+        this.hack_extension = hack_extension;
+        return this;
+    }
+
+    // ------------------------------------------------------------------------- Cache Population
+    // ------------------------------------------------------------------------- Guava Cache
+
+    // separate from stream reading/position
+    private byte[] get(int page_number) throws ExecutionException {
+        int maxPage = (int) (this.contentLength / PAGE_SIZE + 1); // Don'd read past end of stream
+
+        //int firstPage = page_number>9?page_number-10:0; // Get prior cache page, just in case
+        int firstPage = page_number > 0 ? page_number - 1 : 0; // Get prior cache page, just in case
+        int lastPage = (page_number + NUM_PAGES - 1) > maxPage ? maxPage : (page_number + NUM_PAGES - 1);
+
+        for (int i = firstPage; i < lastPage; i++) {
+            final int page_i = i;
+            new Thread(() -> {
```

Comment:
If we create Thread in this way then we won't get the actual impression of thread execution status. You can use executor service & keep track of Future object if needed in this case.

Reply:
Is it OK for me to just create an ExecutorService here when I need it?